### DISCUSSION TITLE
Details about the selection of structures in the dp test command

### DISCUSSION THREAD

**Discussion by vwcruzeiro** [deepmodeling/deepmd-kit#637] [Q&A] [answered]
Hello everyone,

I would like to use the `dp test` command to find out the errors in the energies, forces, and virials for my **full** training and test sets. Let's assume that my training set is located at `data/set.000` and that my test set is at `data/set.001`. When I execute `dp test -m graph.pb -s data -n 100` it is unclear to me from which `set.00?` folder the 100 structures will be picked from. For this reason, I have moved my training set to `training_data/set.000` and my test set to `test_data/set.000`; then, by ensuring that `type.raw` is placed at both `training_data` and `test_data` folders, when I run `dp test` pointing the `-s` flag to these different folders I should select structures only from either the training or test sets, correct?

Assuming what I explained above is correct, my question is more specifically about the `-n` flag. If I want, for example, to evaluate the full test set, I can set `-d` to `test_data` and `-n` equal to the number of structures in the test set. But, will this ensure that all structures are picked? Let me explain what I mean: in case the selection is completely random, a given structure could be picked more than once while another is never picked. Thus, if there something in the code to prevent a given structure to be selected more than once?

Thank you!

**Comment by tuoping** [comment, marked as answer]
![捕获](https://user-images.githubusercontent.com/80671886/118431644-6a3e7980-b709-11eb-9d16-cfd796317d76.PNG)
You can take a look at deepmd/utils/data.py. As you can see, the test_set is not taken randomly, but taken orderly, with tag "shuffle_test" to determine wether to shuffle the order. That is to say, no structures are selected more than once.

**Reply by vwcruzeiro** [reply]
Thank you very much for your reply!
